### PR TITLE
aver: strip module-level effects before injecting test main (planned 0.13)

### PR DESCRIPTION
**Pull request by jasisz**
## Summary

Aver 0.13 (the upcoming release) introduces a module-level `effects [...]` boundary that the type-checker enforces: every function's `! [Effect]` must be covered, or compilation fails.

The bench harness currently asks the LLM for a function only, then strips the LLM's `main()` and injects its own `Console.print(fn(...))` main with `! [Console.print]`. If the LLM declared a narrower module boundary in the original source — including the very common `effects []` for a "pure" module — the injected main violates that boundary and `aver run` fails with an underdeclared-effects type error before any test case runs.

This PR adds `_strip_module_effects(code)` to the runner, called right after `_strip_aver_main(code)`. It removes the module-header `effects [...]` line (inline or multi-line), so the module reverts to legacy / no-boundary mode and the injected main type-checks.

Compatibility:
- **No-op on Aver 0.12 and earlier** — those versions don't recognise `effects [...]` on the module header, so LLMs don't generate it and the strip pass simply doesn't fire.
- **Necessary the moment Aver 0.13 ships** — once the boundary is enforced, any LLM that follows the language docs and declares it will fail bench without this fix. We measured a 50pp regression on tier1 with claude-haiku-4-5 on 0.13-dev when the doc nudge for `effects []` reaches the model.

So merging now is safe: it doesn't change current behaviour, and it lets `aver-bench` continue to work on the day 0.13 lands without a follow-up rush.

## What changed

- `vera_bench/runner.py`: new `_strip_module_effects(code)` helper, wired into the test-main injection path.
- `tests/test_runner.py`: 4 cases on `TestStripModuleEffects` (inline `[]`, inline list, multi-line list, no-op).

## Test plan

- [x] `pytest tests/test_runner.py` — 101 passed locally
- [x] Full suite `pytest tests/` — 489 passed
- [x] `ruff check` + `ruff format --check` clean
- [x] Tier1 sanity with claude-haiku-4-5 on Aver 0.13-dev: 9/10 (matches Aver 0.12 baseline)

## Context

- Linked discussion: this is the bench-side counterpart to a docs change in jasisz/aver where the language now documents `effects [...]` as a first-class module feature
- VB-T1-007 (`safe_modulo`) remains the one tier1 fail on haiku-4.5 in both old and new world — Result-wrap vs raw Int return — unrelated to this change

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive unit tests covering removal of module-level effects declarations across inline, multi-line and edge-case scenarios.

* **Chores**
  * Improved the evaluation/benchmark pipeline to post-process generated code (removing module-level effects and injected mains) for cleaner, consistently type-checked bench outputs, ensuring more reliable and predictable results.

### CODE REVIEW

**Comment by aallan** [comment]
@jasisz Sorry for the delay on getting back to this — the substantive change is reviewed and looks solid (will leave detailed feedback once CI is green), but it can't merge yet because the `dependency-audit` job is failing on an unrelated issue.

The failure is [CVE-2026-3219](https://nvd.nist.gov/vuln/detail/CVE-2026-3219) in pip 26.0.1: `actions/setup-python@v6` bakes that pip version into its toolchain image, and `pip-audit` running inside the runner reports the runner's own pip as vulnerable. Vera hit the same problem at [aallan/vera#537](https://github.com/aallan/vera/issues/537); we've applied the same workaround in vera-bench at [#64](https://github.com/aallan/vera-bench/pull/64) (with [#63](https://github.com/aallan/vera-bench/issues/63) tracking eventual cleanup).

Once #64 lands you'll just need to rebase this branch onto the new main and CI should go green. Happy to do that for you if "allow edits from maintainers" is on, or you can do it yourself — whatever's easier.

Thanks again for getting ahead of the Aver 0.13 boundary change here. The "no-op until 0.13 ships" property of the strip pass is exactly the right way to land forward-compat work, and we want this in before models start picking up the doc nudge for `effects []`.

**Comment by jasisz** [comment]
Thanks for handling the rebase yourself, and for the quick #64 — saved me a context-switch out of Aver 0.14 work, much appreciated.

Pushed CR's whitespace point as `661a0b7`: `_AVER_EFFECTS_OPEN_RE = re.compile(r"^effects\s*\[")` plus a test covering opener variants the parser still accepts (`effects[`, `effects [`, `effects  [`, `effects\t[`). CI should be green once it runs.

Glad the no-op-then-activate property landed well — happy to use the same shape for future Aver-X.Y boundary changes; that way the bench can stay continuously valid for whichever Aver version a model actually emits.

File: vera_bench/runner.py [hunk 1]
```diff
@@ -562,6 +566,67 @@ def _strip_aver_main(code: str) -> str:
...
+            and stripped
+            and not stripped.startswith("//")
+        ):
+            in_module_header = False
+
+        if (
+            in_module_header
+            and indent_len > 0
+            and (stripped.startswith("effects [") or stripped.startswith("effects["))
+        ):
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera_bench/runner.py:627]: @coderabbitai Valid finding — verified: `stripped.startswith("effects [") or stripped.startswith("effects[")` only matches zero or one whitespace character between the keyword and the bracket, missing `effects   [`, `effects\t[`, etc. that Aver's whitespace-insensitive parser still accepts.

Addressed in commit [661a0b7](https://github.com/aallan/vera-bench/commit/661a0b7) with the regex you suggested (`_AVER_EFFECTS_OPEN_RE = re.compile(r"^effects\s*\[")`) plus a parameterized test (`test_strips_arbitrary_whitespace_between_keyword_and_bracket`) covering four opener shapes: `effects[`, `effects [`, `effects  [`, and `effects\t[`. All seven `TestStripModuleEffects` cases pass; full `tests/test_runner.py` suite green.

**Comment by aallan** [comment]
@jasisz Heads-up: I'd just finished writing the same regex fix locally (`_AVER_EFFECTS_OPEN_RE = re.compile(r"^effects\s*\[")` + a parameterized whitespace-tolerance test) when I went to push and discovered you'd already landed [661a0b7](https://github.com/aallan/vera-bench/commit/661a0b7) doing exactly that. Independent convergence — same regex, same test shape, same four opener variants covered.

Your test is slightly better than what I wrote: asserting `"Console.print]" not in header_part` is more precise than asserting `"effects" not in header_part` — yours would catch a hypothetical regression where the strip removes only the keyword but leaves the bracket contents, mine wouldn't. Good instinct on the test target.

Abandoned my local commit, no duplicate push needed. Verified the suite locally on your branch — 104 in `test_runner.py` pass, including the new whitespace-tolerance case. CI should go green; I'll merge once it does and CodeRabbit posts its final pass.

Thanks again for getting ahead of the Aver 0.13 boundary change here. The whole strip pass — including the regex hardening — is the kind of forward-compat work that's worth landing well before the breaking change ships.

**Comment by aallan** [comment]
@coderabbitai @jasisz Pushed [918c3e3](https://github.com/aallan/vera-bench/commit/918c3e3) addressing the bracket-close finding from the duplicate-comments review.

CodeRabbit was right — this isn't really a duplicate, it's the second half of an earlier two-part observation (opener strict-match + closer strict-match) where Szymon's [661a0b7](https://github.com/aallan/vera-bench/commit/661a0b7) addressed the opener but not the closer. The `endswith("]")` check at lines 600 and 625 had a worse-than-no-op failure mode for `effects [Console.print] // pure module`: the opener regex would match, `endswith("]")` would be False (line ends in `e`), and `skip_until_close` would chew through the function body until something happened to end in `]`.

Applied the suggested fix verbatim (`endswith("]")` → `"]" in stripped` in both spots), with a comment explaining the trailing-comment hazard at each call site so the rationale doesn't evaporate.

Two new tests in `TestStripModuleEffects`:
- `test_strips_inline_effects_with_trailing_comment` — the inline variant; explicitly asserts the function body is intact (i.e. NOT eaten by a runaway skip_until_close).
- `test_strips_multiline_effects_with_trailing_comment_on_close` — same hazard on the closing line of a multi-line list.

Both pass; full `tests/test_runner.py` green at 106 cases (104 + 2 new). `ruff check` and `ruff format --check` clean.

@jasisz — sorry to push to your branch a second time today. Same convergence pattern as this morning's regex fix: CodeRabbit re-raised an unaddressed half of its earlier comment, the suggested diff was small and mechanical, and pushing was faster than waiting. If you'd prefer me to comment-and-let-you-push for future iterations on this PR, just say so — happy to switch modes.

**Comment by aallan** [comment]
@jasisz One more touch on the branch: pushed [87a835e](https://github.com/aallan/vera-bench/commit/87a835e) bumping the bench version `0.0.9` → `0.0.10`. Reasoning is that this PR is a methodology change (post-processing on LLM output before compilation) — same input, different output to the compiler — so result files written by the post-merge harness aren't directly comparable to v0.0.9 result files once Aver 0.13 ships and the strip starts firing on real generations.

On Aver 0.9.5/0.12 the strip is a no-op and scores are byte-identical, so the bump is preemptive — but it records the methodology boundary in `bench_version` now, rather than retroactively when the divergence becomes observable.

Files touched:
- `pyproject.toml`: 0.0.9 → 0.0.10
- `CITATION.cff`: version + date-released bumped together (date 2026-04-29)
- `CHANGELOG.md`: new `[0.0.10]` section with a Compatibility note explaining the no-op-until-Aver-0.13 nuance
- `ROADMAP.md`: prepended a v0.0.10 summary line

Verified `importlib.metadata.version("vera-bench") == "0.0.10"` after a fresh `pip install -e .`, full test suite green at 494 cases.

This should be the last push from us — once CI goes green I'll merge and tag v0.0.10.

**Review by aallan — APPROVED** [review]
CI is all green. Ready to merge.
